wordpress-mobile/release-toolkit
# App Size Metrics: add support for Android Universal APKs

This work is part of the App Metrics project [ref: paaHJt-37V-p2] and implements the fastlane actions necessary to send iOS and Android app size metrics [ref: paaHJt-3od-p2#comment-6098] to the [App Metrics server](https://github.com/Automattic/apps-metrics).

## What

This builds on top of https://github.com/wordpress-mobile/release-toolkit/pull/364 and adds support for including "Universal APKs" metrics for Android.

_(I suggest to get #364 merged first before this #365, to keep the diffs of each PR separate and easier to review in isolation)_

## Why

This will mostly be useful if/when we want to send app size metrics for Installable Builds, which are build using `./gradlew assemble<app>JalapenoDebug` and thus produce a single, Universal `.apk` — as opposed to an `.aab` bundle like we do for beta and release builds that we upload to the Play Store)

## How

 - New `universal_apk_path` parameter for the `android_send_app_size_metrics` action
 - The `aab_path` parameter is no longer mandatory, to allow the case where we only pass the universal APK but no `.aab`. It's still required to have at least _one of_ `universal_apk_path` or `aab_path` (if not both) though.
 - Changed a key used in the Android metrics payload to the more explicit `"Optimized File Size"` for R8-optimized APKs (i.e. `file-size` as reported by `apkanalyzer`, as opposed to the `.apk` file size on disk).
    - Hopefully that will help differentiate in particular the Universal APK's file size on disk vs the Universal APK's optimized file size ("optimized" as in "after R8 optimizations and according to `apkanalyzer`", as opposed to "optimized for a specific device architecture", that latter being instead indicated by the `split:` metadata)
 - Moved the keys used in each metrics payload into constants at the top of the action, making them easier to spot and change

## To Test

 - Check that tests pass on this PR (CI green)
 - Check https://github.com/wordpress-mobile/WordPress-Android/pull/16580, which uses this branch to run the action with the Universal APKs built for WP and JP Installable Builds
    - Verify that CI is green on that WPAndroid PR
    - Verify that [the `*-app-size-metrics.json` files in the "Artifacts" tab of the WPAndroid PR's CircleCI job](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Android/31475/workflows/e7f74f3f-4dcc-4cad-b217-ccead5ea8d94/jobs/124137/artifacts) look as expected.
